### PR TITLE
feat: Cache Policy

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This PR is a demo for the cache policy.

**This PR just provide a simple range based cache, I expect to merge this PR first and than improve the cache policy in the following PRs**

OpenDAL introduces a cache policy that allows users to specify the detailed cache logic based on different inputs.

```rust
/// CachePolicy allows user to specify the policy while caching.
#[async_trait]
pub trait CachePolicy: Send + Sync + Debug + 'static {
    fn on_read(
        &self,
        inner: Arc<dyn Accessor>,
        cache: Arc<dyn Accessor>,
        path: &str,
        args: OpRead,
    ) -> BoxFuture<'static, Result<(RpRead, BytesReader)>>;
}
```

By implemented different cache policy on different cache layer, users can compose cache services together:

```rust
operator = operator
    // Cache via user specified cache service.
    .layer(CacheLayer::new(cache_op).with_policy(RemoteCachePolicy::new(1024 * 1024)))
    // Always prepare an in-memory cache.
    .layer(
        CacheLayer::new(Operator::from_env(Scheme::Moka))
        // TODO: allow using different policy on query
        .with_policy(MemoryCachePolicy::new(1024 * 1024)),
    );
```

This example built a two-level cache:

- Databend's internal in-memory cache
- Users specified cache services (fs, redis)

## Plan

I expect to build a basic cache policy that is useful which means:

- **Not affect the result of query (must be correct)**
- **Not big regression compared to query without cache**
- **Much faster than normal query when we already cached!**
- Allow user to control the different cache policy *(maybe later)*
- Allow users settings for cache (per-query level) *(maybe later)*

### CODE REVIEW

**Comment by BohuTANG** [comment]
Let's perf the new cache 🚀

File: src/common/storage/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
+//  Copyright 2021 Datafuse Labs.
```
Comment (zhang2014) [inline src/common/storage/src/cache.rs:1]:
```suggestion
//  Copyright 2022 Datafuse Labs.
```

**Review by PsiACE — APPROVED** [review]

File: src/common/storage/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                            Error::new(
+                                ErrorKind::Unexpected,
+                                "read from underlying storage service",
+                            )
+                            .set_source(err)
+                        })?;
+                        let bs = Bytes::from(bs);
+
+                        // Ignore errors returned by cache services.
+                        let _ = cache
```
Comment (BohuTANG) [inline src/common/storage/src/cache.rs:83]: how to define the cache size of the `cache`?

**Comment by BohuTANG** [comment]
For example:
How do we integrate a two-level cache for fuse engine block data: memory(1GB) and fs(unlimited)?
If the memory cache size is 1GB, if it reaches then evict the datas to fs.

File: src/common/storage/src/cache.rs (hunk 2 above)
Comment (Xuanwo) [reply src/common/storage/src/cache.rs:83]: > how to define the cache size of the `cache`?

Defined by users input. Policy only decides how to use them.

**Comment by Xuanwo** [comment]
> How do we integrate a two-level cache for fuse engine block data: memory(1GB) and fs(unlimited)?
> If the memory cache size is 1GB, if it reaches then evict the datas to fs.

Ideally, it would be like this:

```rust
operator = operator
    .layer(CacheLayer::new(fs_op).with_policy(FuseCachePolicy::new()))
    .layer(CacheLayer::new(memory_op).with_policy(MemoryCachePolicy::with_capcity(1024 * 1024 * 1024)));
```

FuseCachePolicy can share metrics with MemoryCachePolicy to decide which op to take.

**Comment by BohuTANG** [comment]
I am working on benching the new cache.

**Comment by Xuanwo** [comment]
> I am working on benching the new cache.

The new cache is not ready for benching now.

**Comment by BohuTANG** [comment]
## Test Results

This test is benched on this PR, not tier cached(next PR), the result is only for reference Cc @Xuanwo 
You can bench your next tier-cached PR based on this test.

### How to test

[queries.sql](https://raw.githubusercontent.com/ClickHouse/ClickBench/main/databend/queries.sql)

`run.sh`

```
#!/bin/bash

TRIES=3
QUERY_NUM=1

header="| query  | 1st(Cold) | 2nd(Hot) | 3rd(Hot) |\n| ------- | ------- | ------- | ------- |\n";
printf "$header"
cat queries.sql | while read query; do
    echo -n "| q${QUERY_NUM} | "
    for i in $(seq 1 $TRIES); do
        RES=$(curl -w 'Time: %{time_total}\n' http://default@localhost:8124?database=hits -d "${query}" 2>&1 | grep -P '^Time: ' | sed 's/Time: //')
        [[ "$?" == "0" ]] && echo -n "${RES}s" || echo -n "null"
        [[ "$i" != $TRIES ]] && echo -n " | "

        echo "${QUERY_NUM},${i},${RES}" >> result.csv
    done
    echo "|"

    QUERY_NUM=$((QUERY_NUM + 1))
done
```

Run Bench:

```
./run.sh
```





### Only Enable Fuse Meta Cache(Default)

Enable the Fuse Cache:

```
table_cache_enabled = true
```

Disable the block cache:

```
export STORAGE_CACHE_TYPE=none
```

Results:

| query | 1st(Cold)  | 2nd(Hot)   | 3rd(Hot)   |
| ----- | ---------- | ---------- | ---------- |
| q1    | 0.036475s  | 0.034330s  | 0.033860s  |
| q2    | 0.405285s  | 0.425448s  | 0.405614s  |
| q3    | 0.482817s  | 0.505521s  | 0.641434s  |
| q4    | 0.499226s  | 0.565254s  | 0.499185s  |
| q5    | 4.591048s  | 4.641461s  | 4.627463s  |
| q6    | 2.880130s  | 2.768725s  | 2.862527s  |
| q7    | 0.448876s  | 0.403117s  | 0.395856s  |
| q8    | 0.444990s  | 0.415852s  | 0.416349s  |
| q9    | 1.309182s  | 1.233529s  | 1.285408s  |
| q10   | 1.552085s  | 1.507529s  | 1.408535s  |
| q11   | 1.235884s  | 1.146333s  | 1.110571s  |
| q12   | 1.171811s  | 1.448887s  | 1.106544s  |
| q13   | 1.181518s  | 1.121305s  | 1.093806s  |
| q14   | 2.014658s  | 1.989183s  | 1.948704s  |
| q15   | 1.491372s  | 1.468302s  | 1.447846s  |
| q16   | 1.304342s  | 1.374811s  | 1.346403s  |
| q17   | 3.264561s  | 3.240071s  | 3.263775s  |
| q18   | 2.695813s  | 2.634235s  | 2.584193s  |
| q19   | 6.670163s  | 6.639486s  | 6.507481s  |
| q20   | 0.155114s  | 0.121644s  | 0.088172s  |
| q21   | 3.546669s  | 3.656849s  | 3.552669s  |
| q22   | 3.860203s  | 3.845587s  | 4.168806s  |
| q23   | 7.651057s  | 7.943695s  | 7.989370s  |
| q24   | 26.020538s | 22.834948s | 23.258524s |
| q25   | 1.370762s  | 1.356892s  | 1.378027s  |
| q26   | 0.729762s  | 0.815815s  | 0.678436s  |
| q27   | 1.396373s  | 1.543097s  | 1.387461s  |
| q28   | 3.548696s  | 3.508477s  | 3.612818s  |
| q29   | 6.181733s  | 6.261909s  | 6.071892s  |
| q30   | 1.299608s  | 1.234531s  | 1.228019s  |
| q31   | 1.713323s  | 1.645558s  | 1.637541s  |
| q32   | 2.416340s  | 2.529084s  | 2.402624s  |
| q33   | 13.003390s | 11.816549s | 11.737593s |
| q34   | 5.694307s  | 5.897234s  | 5.655931s  |
| q35   | 9.171278s  | 9.242920s  | 9.373954s  |
| q36   | 2.034718s  | 2.029365s  | 2.003341s  |
| q37   | 0.985171s  | 0.732181s  | 0.831386s  |
| q38   | 0.890445s  | 0.419350s  | 0.392929s  |
| q39   | 0.535444s  | 0.528808s  | 0.431822s  |
| q40   | 1.729394s  | 2.310463s  | 2.380718s  |
| q41   | 0.420815s  | 0.466013s  | 0.277656s  |
| q42   | 0.523699s  | 0.373747s  | 0.322487s  |
| q43   | 0.231658s  | 0.349807s  | 0.361606s  |





### Enable Fuse Meta Cache , Enable Block Cache(FS)

Enable the Fuse Cache:

```
table_cache_enabled = true
```

Enable the block cache:

```
export STORAGE_CACHE_TYPE=fs
export STORAGE_CACHE_FS_DATA_PATH=/root/databend/cache_data
```



Results:

| query | 1st(Cold)  | 2nd(Hot)   | 3rd(Hot)   |
| ----- | ---------- | ---------- | ---------- |
| q1    | 0.038899s  | 0.036281s  | 0.035496s  |
| q2    | 0.970526s  | 0.322004s  | 0.317814s  |
| q3    | 0.815075s  | 0.471779s  | 0.394033s  |
| q4    | 1.005818s  | 0.398018s  | 0.335014s  |
| q5    | 4.369559s  | 4.401180s  | 4.389257s  |
| q6    | 3.222771s  | 2.737465s  | 2.797589s  |
| q7    | 0.639833s  | 0.317799s  | 0.317697s  |
| q8    | 0.305690s  | 0.334464s  | 0.318122s  |
| q9    | 1.323162s  | 1.156697s  | 1.143692s  |
| q10   | 1.313659s  | 1.258052s  | 1.223133s  |
| q11   | 1.240450s  | 0.835830s  | 0.787678s  |
| q12   | 0.950553s  | 0.770299s  | 0.732248s  |
| q13   | 0.915870s  | 0.909941s  | 0.924551s  |
| q14   | 1.774458s  | 1.592704s  | 1.552674s  |
| q15   | 1.240069s  | 1.141205s  | 1.077771s  |
| q16   | 1.243651s  | 1.262606s  | 1.198304s  |
| q17   | 3.025028s  | 2.975043s  | 2.886913s  |
| q18   | 2.338710s  | 2.359535s  | 2.366674s  |
| q19   | 6.675166s  | 6.082529s  | 6.173810s  |
| q20   | 0.252197s  | 0.078937s  | 0.088268s  |
| q21   | 20.190976s | 3.114085s  | 3.141822s  |
| q22   | 3.374004s  | 3.411784s  | 3.405602s  |
| q23   | 15.624755s | 6.807229s  | 6.758537s  |
| q24   | 66.900819s | 15.888634s | 15.917083s |
| q25   | 0.809001s  | 0.794073s  | 0.766057s  |
| q26   | 0.477541s  | 0.433566s  | 0.429163s  |
| q27   | 0.758854s  | 0.790309s  | 0.765767s  |
| q28   | 3.250752s  | 3.206791s  | 3.151450s  |
| q29   | 5.980161s  | 5.577301s  | 5.564609s  |
| q30   | 1.148961s  | 1.095830s  | 1.095542s  |
| q31   | 1.398375s  | 1.255473s  | 1.363225s  |
| q32   | 2.245778s  | 2.189994s  | 2.206648s  |
| q33   | 14.728293s | 15.088396s | 15.160674s |
| q34   | 18.635366s | 5.302110s  | 5.122521s  |
| q35   | 8.650739s  | 8.768178s  | 8.916671s  |
| q36   | 1.897953s  | 1.758700s  | 1.761705s  |
| q37   | 0.892157s  | 0.665173s  | 0.659919s  |
| q38   | 0.990978s  | 0.410690s  | 0.431635s  |
| q39   | 0.354378s  | 0.312445s  | 0.316853s  |
| q40   | 1.887725s  | 1.714681s  | 1.719854s  |
| q41   | 0.481858s  | 0.205522s  | 0.186259s  |
| q42   | 0.241733s  | 0.189387s  | 0.340089s  |
| q43   | 0.328778s  | 0.154456s  | 0.218879s  |



The problem is that we write through the block data when read, so the first query is slow if we read lot of data.

File: src/query/storages/fuse/fuse/src/fuse_table.rs [hunk 3]
```diff
@@ -118,10 +118,8 @@ impl FuseTable {
         operator = operator.layer(StorageMetricsLayer::new(data_metrics.clone()));
         // If cache op is valid, layered with ContentCacheLayer.
         if let Some(cache_op) = CacheOperator::instance() {
-            operator = operator.layer(ContentCacheLayer::new(
-                cache_op,
-                ContentCacheStrategy::Fixed(1024 * 1024),
-            ));
+            operator =
+                operator.layer(CacheLayer::new(cache_op).with_policy(FuseCachePolicy::new()));
```
Comment (BohuTANG) [inline src/query/storages/fuse/fuse/src/fuse_table.rs:122]: If the cache is not set:`export STORAGE_CACHE_TYPE=none` (default), does FuseCachePolicy still works?

Comment (Xuanwo) [reply src/query/storages/fuse/fuse/src/fuse_table.rs:122]: > If the cache is not set:`export STORAGE_CACHE_TYPE=none` (default), does FuseCachePolicy still works?

No, it won't work.

**Comment by Xuanwo** [comment]
Unit test always failed for timeout, have no idea on this.

**Comment by BohuTANG** [comment]
test sessions::query_ctx::test_get_storage_accessor_s3 has been running for over 60 seconds

**Comment by Xuanwo** [comment]
> test sessions::query_ctx::test_get_storage_accessor_s3 has been running for over 60 seconds

Can't reproduce locally

**Review by BohuTANG — APPROVED** [review]

**Comment by Xuanwo** [comment]
> Can't reproduce locally, I'm investigating.

I got it. opendal will try to load credential from the environment. Let's disable it in test.

**Pull request merged**